library/src/main/java/com/novoda/downloadmanager/Optional.java

```diff
@@ -0,0 +1,88 @@
+package com.novoda.downloadmanager;
+
+public final class Optional<T> {
```

Comment:
How can we avoid this `Optional` class to be leaked to the client libraries?

With this addition, client's will be able to just use this class since it is public and exposed. 

First of all, If they have java8, then it will be a problem since it is redundant. 
Also they may already have `Optional` class in their project. So this will be a duplication in auto completions in IDE and can be annoying. 

**Possible Solutions:** 
- Create another gradle module called `internal` and move this there and depend on `internal` by using `implementation`. Gradle hides it for you. 
- Or if that's complicated, simply move this to `internal` package. (OkHttp does this). This way it will be little bit clear for library users that they are not supposed to use this.

Reply:
Also another problem is that, in the future where we increase our minSdk to suppor java 8 natively, it will be hard to remove this class because clients may use this and since it is not marked as internal, just removing this will be really annoying for them.

Reply:
9351e07